### PR TITLE
feat: add sendMouse command

### PR DESCRIPTION
**Pull request by vursen**
## What I did

- [x] Introduced a new command `sendMouse`.
- [x] Added basic unit tests.
- [x] Added support for Playwright.
- [x] Added support for Puppeteer.
- [x] Added support for WebDriver.
- [x] Make unit tests work in Playwright Webkit.
- [x] Make unit tests work in Windows Chrome.

### Move
```ts
function sendMouse(payload: {
  type: 'move',
  position: [number, number]
}): Promise<void>;
```

### Click
```ts
function sendMouse(payload: {
  type: 'click',
  position: [number, number],
  button?: 'left' | 'middle' | 'right'
}): Promise<void>;
```

### Down
```ts
function sendMouse(payload: {
  type: 'down',
  button?: 'left' | 'middle' | 'right'
}): Promise<void>;
```

### Up
```ts
function sendMouse(payload: {
  type: 'up',
  button?: 'left' | 'middle' | 'right'
}): Promise<void>;
```

## Future improvements

It is possible to extend the command's API in the future to also support sending a sequence of mouse actions, e.g.:

```js
await sendMouse([
  { type: 'move', position: [100, 100] },
  { type: 'move', position: [200, 200] }
]);
```

### CODE REVIEW

File: packages/test-runner-commands/test/send-mouse/sendMousePlugin.test.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  it('can send mouse on puppeteer', async () => {
+    await runTests({
+      files: [path.join(__dirname, 'browser-test.js')],
+      browsers: [chromeLauncher()],
+      plugins: [sendMousePlugin()],
+    });
+  });
+
+  // playwright doesn't work on windows VM right now
+  if (platform() !== 'win32') {
```
Comment (web-padawan) [inline packages/test-runner-commands/test/send-mouse/sendMousePlugin.test.ts:36]: You can remove this check, most of Playwright tests have been passing lately.

File: packages/test-runner-commands/test/send-mouse/sendMousePlugin.test.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import { chromeLauncher } from '@web/test-runner-chrome';
+import { webdriverLauncher } from '@web/test-runner-webdriver';
+import { playwrightLauncher } from '@web/test-runner-playwright';
+import { sendMousePlugin } from '../../src/sendMousePlugin';
+import { startSeleniumServer } from '../selenium-server';
+
+let seleniumServer: selenium.ChildProcess;
+
+describe('sendMousePlugin', function test() {
+  this.timeout(20000);
```
Comment (web-padawan) [inline packages/test-runner-commands/test/send-mouse/sendMousePlugin.test.ts:14]: Let's try the same timeout as `webdriver` launcher (and `selenium` launcher) tests.

```suggestion
  this.timeout(50000);
```

File: packages/test-runner-webdriver/src/IFrameManager.ts [hunk 3]
```diff
@@ -155,6 +155,11 @@ export class IFrameManager {
     return this.driver.keys(keys);
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async performActions(_: string, _actions: object[]) {
+    console.warn('Device actions are not supported in the concurrent mode.');
```
Comment (web-padawan) [inline packages/test-runner-webdriver/src/IFrameManager.ts:160]: I would rather explicitly mention the necessary config option, e.g. something like this:

```
Unsupported operation. In order to use device actions in Webdriver, set "concurrency" to 1.
```

File: packages/test-runner-commands/test/send-mouse/sendMousePlugin.test.ts (hunk 1 above)
Comment (vursen) [reply packages/test-runner-commands/test/send-mouse/sendMousePlugin.test.ts]: Done.

File: packages/test-runner-commands/test/send-mouse/sendMousePlugin.test.ts (hunk 2 above)
Comment (vursen) [reply packages/test-runner-commands/test/send-mouse/sendMousePlugin.test.ts]: Done.

File: packages/test-runner-webdriver/src/IFrameManager.ts (hunk 3 above)
Comment (vursen) [reply packages/test-runner-webdriver/src/IFrameManager.ts]: Good point, done.

File: packages/test-runner-commands/src/sendMousePlugin.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      `You must provide a type option with one of the following values: ${validTypes.join(', ')}.`,
+    );
+  }
+
+  if (['click', 'move'].includes(payload.type)) {
+    if (
+      !Array.isArray(payload.position) ||
+      typeof payload.position[0] !== 'number' ||
+      typeof payload.position[1] !== 'number' ||
+      !/^\d+$/.test(payload.position.join(''))
```
Comment (web-padawan) [inline packages/test-runner-commands/src/sendMousePlugin.ts:55]: Consider using [`Number.isInteger`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger) for better readability.

File: packages/test-runner-webdriver/src/IFrameManager.ts [hunk 5]
```diff
@@ -155,6 +155,13 @@ export class IFrameManager {
     return this.driver.keys(keys);
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async performActions(_: string, _actions: object[]) {
+    console.warn(
+      'Unsupported operation. In order to use device actions in Webdriver, set "concurrency" to 1.',
+    );
```
Comment (web-padawan) [inline packages/test-runner-webdriver/src/IFrameManager.ts:162]: I think `console.warn` isn't quite right here. Should we throw an error instead?

**Review by web-padawan — CHANGES_REQUESTED** [review]
For some reason tests fail on Windows, need to investigate what goes wrong there 🤔 

```
  1) sendMousePlugin
       can send mouse on puppeteer:
     Error: Test run did not pass
      at EventEmitter.<anonymous> (packages\test-runner-core\src\test-helpers.ts:143:16)
      at EventEmitter.emit (domain.js:483:12)
      at TestRunner.emit (packages\test-runner-core\src\utils\EventEmitter.ts:24:20)
      at TestRunner.stop (packages\test-runner-core\src\runner\TestRunner.ts:178:10)

  2) sendMousePlugin
       can send mouse on playwright:
     Error: Test run did not pass
      at EventEmitter.<anonymous> (packages\test-runner-core\src\test-helpers.ts:143:16)
      at EventEmitter.emit (domain.js:483:12)
      at TestRunner.emit (packages\test-runner-core\src\utils\EventEmitter.ts:24:20)
      at TestRunner.stop (packages\test-runner-core\src\runner\TestRunner.ts:178:10)

  3) sendMousePlugin
       can send mouse on webdriver:
     Error: Test run did not pass
      at EventEmitter.<anonymous> (packages\test-runner-core\src\test-helpers.ts:143:16)
      at EventEmitter.emit (domain.js:483:12)
      at TestRunner.emit (packages\test-runner-core\src\utils\EventEmitter.ts:24:20)
      at TestRunner.stop (packages\test-runner-core\src\runner\TestRunner.ts:178:10)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by Westbrook** [comment]
One thing we caught with this shape of API recently in Spectrum Web Components is that if you have multiple tests in the same file and you pointer down in one test the pointer won't be reset for the following test. To support that in the equivalent of `commands.mjs` we needed to add this: https://github.com/adobe/spectrum-web-components/blob/main/test/plugins/browser.ts We don't test the plugin, but it caused a failure in [this test](https://github.com/adobe/spectrum-web-components/blob/main/packages/slider/test/slider.test.ts#L488) because this [earlier test](https://github.com/adobe/spectrum-web-components/blob/d362050c2fd2c8b105bb32929923cfe69ddbb1b3/packages/slider/test/slider.test.ts#L334-L372) wasn't previously releasing the pointer.

File: packages/test-runner-commands/src/sendMousePlugin.ts (hunk 4 above)
Comment (vursen) [reply packages/test-runner-commands/src/sendMousePlugin.ts]: Done.

File: packages/test-runner-webdriver/src/IFrameManager.ts (hunk 5 above)
Comment (vursen) [reply packages/test-runner-webdriver/src/IFrameManager.ts]: Done.

**Comment by vursen** [comment]
Yeah, that's not an obvious thing. I ran into that once while working on the plugin unit tests: it was not possible to `down` any mouse button until the previous one is not `up` and for that reason, I test `down` and `up` within one test case rather than within separate ones. Nevertheless, doesn't the current `sendKeys` API have the same problem?

```js
await sendKeys({ down: 'Shift' });
// ...
await sendKeys({ up: 'Shift' }); // <-------- should be explicitly called to release a key
```

I wonder if there is actually a way to get it around or make it somehow work more obvious for the user other than just mention it in the docs e.g:  "remember to release the mouse button after you down it"?

At least, we might consider auto-releasing the previously pressed mouse button once the user sends `down` for another button.

```
await sendMouse({ type: 'down' });
await sendMouse({ type: 'down' }); // <----- will release the left mouse button before making it `down` again.
```

...or maybe having a built-in wrapper around `sendMouse` with a queue like you shared would be even more preferable.

**Comment by daKmoR** [comment]
how about a generic "reset function"?

e.g.  something like this

```js
await sendMouse({ type: 'down' });

// do some expects

await resetMouse(); // move to 0 0 (or another location?) and don't press any button

```

a similar approach could work for the keys as well?

then you can run it as a cleanup/setup as afterEach/beforeEach

**Comment by web-padawan** [comment]
Personally, I like the idea about `resetMouse()` command - and IMO it's better to call it manually in `adterEach`.

**Comment by Westbrook** [comment]
I like a reset method, but I don't like not automatically calling it in the after/teardown phase. Having run into the error, it takes forever to debug the issue there and we shouldn't be placing traps in our API for users to fall into like that.

**Comment by daKmoR** [comment]
where/how would you add it? the `sendMouse` can not automatically add something to the afterEach? I mean we can add it to our docs and to our default setups but I don't think we can do more? at least I don't know of a way 🤔

**Comment by vursen** [comment]
I'm not 100% sure if it is a good idea to reset the mouse (or the keyboard) on `afterEach`. Consider the following examples:

**sendMouse API:**
```js
describe('mouse', () => {
  before(async () => {
    await sendMouse({ type: 'move', position: [100, 100] }); 
  });

  it('should click once', () => {
    await sendMouse({ type: 'up' }); 
    await sendMouse({ type: 'down' }); 
  });

  it('should click once again', () => {
    // The mouse position will be reset to 0, 0 which is not obvious.
    await sendMouse({ type: 'up' }); 
    await sendMouse({ type: 'down' }); 
  });
});
```

**sendKeys API:**
```js
describe('keyboard', () => {
  before(async () => {
    await sendKeys({ down: 'Shift' });
  });

  after(async () => {
     await sendKeys({ up: 'Shift' }); 
  });

  it('should tab once', () => {
    await sendKeys({ press: 'Tab' });
  });

  it('should tab once again', () => {
    // The Shift key is not down anymore which is not obvious.
    await sendKeys({ press: 'Tab' });
  });
});
```

I don't see anything that would stop the user from doing this since it looks like a legal way to use the commands.

**Comment by web-padawan** [comment]
I agree with @vursen and generally I would object enabling side-effects by default. Let's make sure we document the reset method. I would like to proceed with this PR once the build is green.

**Comment by daKmoR** [comment]
I personally do not "like" these "global" setups... personally, I prefer each it as it's own individual unit... 

why I think it is bad

```js
describe('el', () => {
  let obj;
  before(async () => {
    obj = await setupObj(); // returns { foo: 'fooValue' }
  });

  it('first', () => {
    obj.bar = true; // we set this for something else we test
    expect(obj.foo).to.equal('fooValue');
  });
  
  it('second', () => {
    expect(obj.bar).to.be.true;
    // this now succeeds if run after "first" even though it should fail
  });
});
```


that is also how it's currently handled for fixture of  `@open-wc/testing`.

```js
import { fixture } from '@open-wc/testing';

describe('el', () => {
  let el;
  before(async () => {
    el = await fixture(html`<my-el></my-el>`);
  });

  it('first', () => {
    expect(el)...
  });
  // fixture get's clean up in afterEach

  it('second', () => {
    expect(el)... // this will error now
  });
});
```

imho if you set something up then it should be done in a `beforeEach` so every `it()` has a clean slate.

there are probably situations where sharing these "states" across `its` makes sense... but then it's the responsibility of the author to create and account for that.

In short
- we should document how to use `resetMouse` via `afterEach`
- if we add it to `@open-wc/testing` then we should add it to its "global" `afterEach`
- we should not apply any "magic" to automatically clean itself up
- (this recommends a good practice to use beforeEach & afterEach but does not prevent different uses like with shared states)

**Comment by web-padawan** [comment]
@daKmoR Thank you for putting this together! I fully agree with your proposal, so we are on the same line here 💯

**Comment by Westbrook** [comment]
@vursen the decision to write a test as you've supplied is a decision not to be making a unit test, as you are not cleanly separating your units as preparing yourself for any number of unknown issues from the bleed of test data from one test to the next.

What's more, if you were to add a test as follows to the end, it would fail/not test the interaction you were intending for not being able to complete the `click` action (at least in Playwright) as the mouse will already be _down_ and the internal API is to send a `down` then an `up` which means you'll only be able to register the `up` command: 

```
  it('should click once again', () => {
    // The mouse position will be reset to 0, 0 which is not obvious.
    await sendMouse({ type: 'click' }); 
  });
```

I'd very much say that we are similarly letting consumers down with the `sendKeys` API as it stands, but I'm not sure why it was easier to run into this issue with mouse in our actual usage.

> if we add it to @open-wc/testing then we should add it to its "global" afterEach

Why would we want our web component consumers to be better covered in this situation that our "web" consumers?

If at large the team is against cleaning up by default, I'd like to have clean up added optionally to the API signature. If we weren't batching the whole command at the options level, I'd propose:
```
await sendMouse({
   steps: [
        {type: 'click' }
    ],
    cleanup: true/false
); 
```
But seeing as we are, maybe a second argument?
```
const cleanup = true; /// false by default
await sendMouse({ type: 'click' }, cleanup); 
```

**Comment by daKmoR** [comment]
`@web` does not have any testing helpers for specific frameworks so far right? 

so imho sendMouse should not be aware of the testing framework it is used on... that cleanup would fail if you use it with uvu or  any other testing system that is not using afterEach...

if you really need that then we could expose it as a wrapping function like `sendMouseBDDCleanup` or so... but it should only be a small section in the docs... most of the example should be about sendMouse

alternatively, we could introduce a `@web/testing` package and bundle it there... but I'm not sure if that make it even more confusing 🙈

**Comment by Westbrook** [comment]
These are decent options. I do think we need to ship one from the start. 

The `sendMouseBDDCleanup` idea kind of implies a middle ground of a `bindSendMouseWithCleanup()` helper (needing lots of bikeshedding on naming) that you configure at the beginning of your tests that would allow for this to managed more cleanly for you. In that way `uvu` via `asdfg` might look like:

```
import { describe, executeTests } from '@asdgf/core';
import { expect } from '@open-wc/testing';
import { bindSendMouseWithCleanup } from '@web/test-runner-commands';

describe('button', ({it, beforeEach, afterEach }) => {
  const sendMouse = bindSendMouseWithCleanup(afterEach);
  let button;
  let count;

  beforeEach(() => {
    button = document.createElement('button');
    count = 0;
    button.addEventListener('click', () => count += 1);
    button.addEventListener('dragstart', () => count += 1);
  });

  it('drags', () => {
    await sendMouse({type: 'down'});
    await sendMouse({type: 'move', position: [100, 100]);
    expect(count)).to.equal(1);
  });

  it('clicks', () => {
    sendMouse({ type: 'click' });
    expect(count)).to.equal(1);
  });
});
```

This protects you both from the foot gun on not cleaning up at all, while also preventing multiple cleanups via a naive wrapping of the command locally that can overrun the test timeout, which is why you see the use of `mouseCleanupQueued` in the version I currently use: https://github.com/adobe/spectrum-web-components/blob/main/test/plugins/browser.ts#L57

**Comment by web-padawan** [comment]
@Westbrook Please note that with this discussion we are probably setting the bar too high for a first time contributor.

Would it be an option to implement `sendMouse` first, and then add the `reset` / `cleanup` API in a subsequent PR?
Personally, I would name it `releaseMouse()` and then call it manually in `afterEach`. That's simple and clean.

BTW, we can move fast and break things, as this project is still at 0.x, so we can change the API any time 🙂

**Review by web-padawan — APPROVED** [review]
LGTM. 

Two things that IMO can be done in separate PRs:
1. Add documentation for the new command [here](https://github.com/modernweb-dev/web/blob/master/docs/docs/test-runner/commands.md),
2. Implement `resetMouse` as we previously discussed.
